Feature/netstandard2 (NETCore 2.1 support)

kll opened the pull request:
Changing the target to netstandard2.0 allows upgrading to the latested versions of LibGit2Sharp and its associated LibGit2Sharp.NativeBinaries packages.

This allows the dotnetcore version to run on all the [.NET Core 2.1 supported OS](https://github.com/dotnet/core/blob/master/release-notes/2.1/2.1-supported-os.md) versions. 

I left the unit tests targeting the full desktop framework but bumped them to net461 so they could consume the netstandard2.0 GitVersionCore project. The reason I didn't convert everything over is that there are several nuget packages used in the test projects that haven't been updated in a long time and do not have any netstandard support at all.

li0nsar3c00l commented:
The mentioned PR has been merged and a preview of LibGit2Sharp has been published to [nuget](https://www.nuget.org/packages/LibGit2Sharp/0.26.0-preview-0027). When can we expect a merge?

kll commented:
@li0nsar3c00l  I can't really say, PRs seem to hang around for really long times in this repo. :smile:  The PR this one is based on has been around for almost 11 monts now. Looks like it is finally about to merge though.

kll commented:
@JakeGinnivan @dazinator There was discussion in #1269 about merging in GitTools.Core and I agree with the idea. Do you all want me to do that as part of this PR? It would certainly simply things not having to git the PR in GitTools.Core merged/released first.

JakeGinnivan commented:
@kll I recon. Lets do it, anything to simplify this project for now.

The plan was to have a useful separate testing library, but we can release it from here if people want it. I rebased this PR to remove the duplicated commits, it's running for me so if you want to move those two libraries into this PR that is cool

kll commented:
@JakeGinnivan Done.

I also went ahead and updated to the latest prerelease of LibGit2Sharp so that we could use the latest LibGit2Sharp.NativeBinaries which solves my primary problem of being able to easily use GitVersion from the latest dotnet docker images. I had it referencing a release version of LibGit2Sharp but since GitTools.Core was previously referencing a prerelease version of LibGit2Sharp anyway I decided to continue the trend for now.

li0nsar3c00l commented:
@kll thanks for all your effort! Hoping for a fast merge.

dazinator commented:
Does this overlap with #1416 and #1248 ? Can those be closed in favour if this, or vice versa?

li0nsar3c00l commented:
@dazinator Imo yes, #1416 and #1248 can both be closed since this pullrequest includes the latest preview version

dazinator commented:
Before we merge and release this to nuget.org, can we ensure we release whats currently in master to nuget.org, as those packages will support pre `netstandard20` consumers, and then after this change, netstandard20 will become the new minimum.

gep13 commented:
@JakeGinnivan what do you think about this?  Should we release a 4.0.0 that supports netstandard 1.3 and then another release 4.1.0 that supports 2.0?  Does that open up a lot of potential for support issues?  @dazinator what are your thoughts here?

dazinator commented:
@gep13 - That would be my preferred route yes. The PR that targets the lower `netstandard` version is needed for anyone that has projects stuck on the `net core sdk 1.1.x` tool chain. This includes several of my own projects - I can't update these projects yet to use the .net sdk 2.0 or 2.1 toolchain yet - because of breaking issues like this: https://github.com/dotnet/cli/issues/5331 

If we don't release this to nuget.org, I can always continue to consume the appveyor ci nuget feed to obtain the PR packages, as long as those packages are retained. However I would prefer a proper release to nuget.org for peace of mind because I am not sure what the retention rules are in appveyor for packages and pr branches.

Once this PR is merged, we should just say we no longer support the older version, _but atleast it's available_ for those that need it like me :-)

li0nsar3c00l commented:
To ensure compatibility with higher sdk's, I'd recommend the removal of the `src/global.json` file. Apart from that I am wondering, why the unit tests for `GitTools.Testing` uses `xUnit` while all the other projects use `NUnit`. In Addition, some nuget dependencies could be updated.
Can you guys tell me, what is missing for merging this PR?

JakeGinnivan commented:
@dazinator isn't part of the reason .net core is good is you can side by side. Why can't we land this and then external projects still be on 1.1? (honest question)

dazinator commented on src/GitVersionTask/GitVersionTask.csproj:
```diff
@@ -9,7 +9,7 @@
     <RootNamespace>GitVersionTask</RootNamespace>
     <AssemblyName>GitVersionTask</AssemblyName>
-    <TargetFrameworks>net461;netstandard1.5</TargetFrameworks>
+    <TargetFrameworks>net461;netstandard2.0</TargetFrameworks>
```
Can `net461` now be removed here, as net461 implements netstandard20?

kll replied:
Yeah I wasn't really sure if it was OK to remove it in the task project or not because I didn't have any real experience with it so I decided it was safer to leave it alone until I was told otherwise. :)

li0nsar3c00l replied:
As far as I know, that way it can be referenced from other .NET Standard libs and doesn't cause problems with an implicit reference to the full .NET Framework

kll replied:
@li0nsar3c00l That isn't how referencing multi-targeted nuget packages works. The project system will do the right thing and only pull in the most applicable version of the referenced library. If you reference it from a netstandard2 project you will only end up with a reference to the netstandard2 version of the library.

dazinator commented:
@li0nsar3c00l 
I can't see any obvious blockers for merging this.
Just to be clear this advances some platform baselines:

- GitVersionExe: `net40` --> `net461` (and NET Core 2.0)
-`GitVersionTask`: `netstandard1.5` --> `netstandard2.0`.

I think it would be good to get a release of master out the door for reasons mentioned above, and also, at the moment I don't see any evidence that we have a handle on releasing. It would be good to get out house in order on the releasing front before we accept any future PR's however, we could merge this first and then sort the release. @JakeGinnivan  @gep13  what say you?

li0nsar3c00l commented:
@kll Any thoughts on my recent comment? Do you agree, that this should be fixed? In case you do, should I create a pull request against your repo or do you want to edit that on your own?

petriashev commented:
Anybody PRESS MERGE BUTTON!!! )

dazinator commented:
@petriashev merging this won't do much good until we also sort #1445 - because merging wont result in any release at present.

derwasp commented:
I am not sure whether it's intended or not, but when `GitVersionCore` is created, it depends on `GitTools.Core`, which is not built using the script. Just my 5 cents ;)

youfoundkris commented:
Does appveyor have a nuget repository I can access to already start using this?

dazinator commented:
@youfoundkris - I appreciate the sentiment. Having spent a long time porting stuff to netstandard and netcore, I too want this released as much as anyone! If you can't wait for #1445 there is nothing stopping you from building this PR yourself to get the packages you need in the interim. I think we would all prefer to have a CI / CD pipeline in place though.

kll commented:
I rebased against the latest from master to fix the conflicts.

kll commented:
@youfoundkris The appveyor feed you linked to is definitely not empty. You have to query it properly to get a list of packages. The easiest way to see what is on there is using the nuget command line. Try this and you'll see its actually got a confusingly large amount of stuff on it.
` nuget list -Source https://ci.appveyor.com/nuget/gitversion-8nigugxjftrw -Prerelease -All`

Add it as a nuget source and you can install whatever version you need. The latest build of this PR from my rebase today is version `4.0.0-pullrequest1422-1622`

kll commented:
This now also includes @arturcic changes to publish it as a dotnet core global tool. You can give it a try by installing it with this command:
` dotnet tool install -g --version 4.0.0-pullrequest1422-1625 --add-source https://ci.appveyor.com/nuget/gitversion-8nigugxjftrw GitVersion.CommandLine.DotNetCore.Tool`

dazinator commented:
I'm looking forward to seeing this merged - great work!

arturcic commented:
@dazinator I think I should first check/fix the https://github.com/GitTools/GitVersion/issues/1445, and after we have the artifacts published, we cand try merging this one. Just to have a version of GitVersion targeting older version of .net core

kll commented:
This is now freshly rebased off of master to take all the build script changes into account and clean up the commits a bit. I'm not sure if there is anything I need to do related to the new build for the global tool that is part of this branch.  I'm having some issue that seems local to my machine causing the build to fail in the Pack-Tfs step which makes it hard for me to fully verify the builds locally. Looking at the AppVeyor build log it passes that step fine but fails in the Teardown step but still reports the build as successful.

arturcic commented:
> the AppVeyor build log it passes that step fine but fails in the Teardown step but still reports the build as successful.

I know about this issue, I'll have a look, seems like it's Appveyor related.

> I'm not sure if there is anything I need to do related to the new build for the global tool that is part of this branch. 

I'll have a look and update the scripts and will send you a PR

arturcic commented:
@kll could you please rebase against master again, as I have updated the scripts and renamed run.ps1 and run.cake to build.ps1 and build.cake. After the rebase I'll update the build scripts for the global tool, and send you a PR

kll commented:
The travis builds are failing in Copy-Files while doing the ILMerge stuff under mono. The specific error is:
`Mono.Cecil.AssemblyResolutionException: Failed to resolve assembly: 'netstandard, Version=2.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'

From my (very limited) knowledge ILMerge doesn't even support mono, right? So shouldn't it just be skipped when building with mono?

kll commented:
I have a really good skill of missing your messages @arturcic :) I was fixing it on my own anyway but decided to be lazy and do a normal merge rather than a rebase. At this point I just want to wrap this up and have stopped caring about how clean the commit history is.

arturcic commented:
when it will be ready the PR, you could squash the commits and make the history look nicer ;)

dazinator commented on README.md:
```diff
@@ -1,24 +1,28 @@
-|               **Docker**   |   [gittools/gitversion][dockerhub]        |                      -                    |
 
 ## Compatibility
 GitVersion works on Mac, Linux with Mono and Windows.
```
.. as well as dotnet core platforms?

arturcic replied:
good catch

kll commented:
So any ideas about what I should do for the travis builds?

arturcic commented:
> So any ideas about what I should do for the travis builds?

it looks like it's a ilrepack issue https://github.com/gluck/il-repack/issues/203

jmezach commented:
So now that #1445 is sorted, perhaps now would be a good time to pick this up again? Does anybody what remains to be done on this?

kll commented:
travis is working now, but the new Azure DevOps build is broken and I'm reasonably certain it has nothing to do with the changes in this PR

arturcic commented:
> travis is working now, but the new Azure DevOps build is broken and I'm reasonably certain it has nothing to do with the changes in this PR

I'll have a look

kll commented:
I'm also still having issues with the global tool packaging. It seems to work fine when I install it directly out of the artifacts folder from a local build but when I try to install the latest version the appveyor feed it fails to even install.
```
$ dotnet tool install -g GitVersion.CommandLine.DotNetCore.Tool --add-source https://ci.appveyor.com/nuget/gitversion-8nigugxjftrw --version 4.0.1-PullRequest1422-0016
The settings file in the tool's NuGet package is invalid: Failed to retrieve tool configuration: Could not find a part of the path '/home/work/.dotnet/tools/.store/gitversion.commandline.dotnetcore.tool/4.0.1-pullrequest1422-0016/gitversion.commandline.dotnetcore.tool/4.0.1-PullRequest1422-0016/tools/netcoreapp2.1/any/DotnetToolSettings.xml'.
Tool 'gitversion.commandline.dotnetcore.tool' failed to install. Please contact the tool author for assistance.
```

The entire main directory under `~/.dotnet/tools/.store/gitversion.commandline.dotnetcore.tool/` is missing so I'm not even sure where to start.

arturcic commented:
> I'm also still having issues with the global tool packaging. It seems to work fine when I install it directly out of the artifacts folder from a local build but when I try to install the latest version the appveyor feed it fails to even install.
> 
> ```
> $ dotnet tool install -g GitVersion.CommandLine.DotNetCore.Tool --add-source https://ci.appveyor.com/nuget/gitversion-8nigugxjftrw --version 4.0.1-PullRequest1422-0016
> The settings file in the tool's NuGet package is invalid: Failed to retrieve tool configuration: Could not find a part of the path '/home/work/.dotnet/tools/.store/gitversion.commandline.dotnetcore.tool/4.0.1-pullrequest1422-0016/gitversion.commandline.dotnetcore.tool/4.0.1-PullRequest1422-0016/tools/netcoreapp2.1/any/DotnetToolSettings.xml'.
> Tool 'gitversion.commandline.dotnetcore.tool' failed to install. Please contact the tool author for assistance.
> ```
> The entire main directory under `~/.dotnet/tools/.store/gitversion.commandline.dotnetcore.tool/` is missing so I'm not even sure where to start.

Are you running it on ubuntu 18.04?

arturcic commented:
https://github.com/cake-build/cake/issues/2282

kll commented:
Arch linux actually. That link was helpful though because it made me see something I had missed before. The previous version that installs is `4.0.0-pullrequest1422-1632` and the new one that won't install is `4.0.1-PullRequest1422-0016` which has capitals in the prerelease tag when before it did not. It definitely seems like some kind of bug in the CLI but I'm going to see if I can work around it by making it all lower case again. I wonder what changed...

arturcic commented:
Seems like it's a cli issue, and it was kind of fixed on version 2.1.5xx of sdk https://github.com/dotnet/cli/issues/10101

arturcic commented:
I have disabled Azure DevOps integration, on next commit it will not be required

kll commented:
All the artifacts created locally have all lower case prerelease tags so I'm not sure where to even start on that work around. It seems to just be the artifacts that are produced in the appveyor build that get the capitalization and I'm not sure what changed recently to cause that.

So I think I'm just going to leave it as-is and hope it gets merged in finally.

arturcic commented:
Finally all check pass

arturcic commented:
@dazinator, @gep13 , @JakeGinnivan  are you ok merging this PR?

devlead commented:
I think you should multi target as .NET standard doesn't work smoothly on all .NET FX versions and I assume GitVersion.exe should support same platforms as it does now?

There's a new libgit2sharp version in the works which will target both netfx and .net standard.

arturcic commented:
> There's a new libgit2sharp version in the works which will target both netfx and .net standard.

@devlead Do you have a PR or issue we can track?

devlead commented:
PR is merged 12 days ago, so hopefully in next release.

li0nsar3c00l commented:
Are you talking about the mono runtime or mono.cecil? The later can easily be replaced.

dazinator commented:
According to latest comments here: https://github.com/GitTools/GitVersion/issues/1503#issuecomment-438621965 GitVersionTask v4.0.0 was already working with dotnet core sdk 2.1.0.. I am a bit confused because I thought it wasn't meant to work on 2.1.0 without this PR!?

kll commented:
I don't really know much about the msbuild task or how it works. I only use the commandline (or now the tool version) exclusively.

And on the subject of not really knowing much, I have no idea what is gone wrong with this PR now. All I did was merge the latest from master and now the test step in the build script just hangs forever. Everything compiles fine and all the tests run and pass in Visual Studio. It is only running them from the build script that hangs.

arturcic commented:
Need to check that

kll commented:
I tried some random stuff like re-doing the merge but as a rebase, consolidating and updating testing related packages and nothing seemed to help. So I went ahead and updated to a prerelease version of libgit2sharp that supports the multi-targeting and added net461 to GitTools.Core and GitVersionCore and then the tests started running again from the build script. So it's working locally for me now at least, lets see if the build servers agree.

arturcic commented:
I have checked it locally as well. The reason it was hanging is because of the code coverage tool that runs the tests (Coverlet).  On master branch it was working fine, collecting the code coverage, and with the changes probably you fixed the issue

arturcic commented:
seems like you fixed the issue

kll commented:
It also now has the requested multi-targeting with net461 and netstandard2 so perhaps maybe we can get this merged soon?

arturcic commented:
I guess this is more about 
> Changing the target to netstandard2.0 allows upgrading to the latested versions of LibGit2Sharp and its associated LibGit2Sharp.NativeBinaries packages.

Meaning we'll be able to update to latest LibGit2Sharp as soon as they appear. @kll, am I wrong?

kll commented:
Correct, in my opinion that is what it is all about. For me specifically it is all about using the newer LibGit2Sharp version which works with the official MS provided Docker images which is what I need for my workflow. Right now it just blows up, including the new tool version already being published to the official NuGet feed. If you install it on most modern versions of Linux or in the Docker images, it just crashes. So this PR is pretty important to get GitVersion working cross-platform.

dazinator commented:
@kll Understood, thanks for the clarification, I think the claims I have seen of things working under 2.1.0 already, must be on "some" platform only, where as this should expand that greatly, and allow us to keep up to date with libgit2sharp as it evolves.

Ok cheers ;-)

jacksondaw commented:
GitVersionTask 4.0.0+ is still broken on microsoft/dotnet:2.1-sdk. Any project with it throws: 
MSBUILD : Task factory error NMSBT001: Exception in initialization: System.MissingMethodException: Method not found: 'Void NuGet.Commands.RestoreRequest..ctor(NuGet.ProjectModel.PackageSpec, NuGet.Commands.RestoreCommandProviders, NuGet.Protocol.Core.Types.SourceCacheContext, NuGet.Common.ILogger)'. [...]
MSBUILD : Task factory error NMSBT001:    at UtilPack.NuGet.BoundRestoreCommandUser.PerformRestore(ValueTuple`2[] targets, CancellationToken token) [...]
MSBUILD : Task factory error NMSBT001:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) [...]
MSBUILD : Task factory error NMSBT001:    at UtilPack.NuGet.BoundRestoreCommandUser.PerformRestore(ValueTuple`2[] targets, CancellationToken token) [...]
MSBUILD : Task factory error NMSBT001:    at UtilPack.NuGet.BoundRestoreCommandUser.RestoreIfNeeded(CancellationToken token, ValueTuple`2[] packageInfo) [...j]
MSBUILD : Task factory error NMSBT001:    at UtilPack.NuGet.MSBuild.NuGetTaskRunnerFactory.Initialize(String taskName, IDictionary`2 parameterGroup, String taskBody, IBuildEngine taskFactoryLoggingHost) [...]

I have tried the latest beta releases as well.

jacksondaw commented:
I guess, I am confused. I have tried multiple lower versions of dotnet sdk, GitVersionTask and UtilPack.NuGet.MSBuild, and I have not found a combination that works.

Issue #1503 says to use microsoft/dotnet:2.1.403-sdk, GitVersionTask 4.0.1-beta* and UtilPack.NuGet.MSBuild 2.0.9.

With that combination I am getting:
 error MSB4175: The task factory "UtilPack.NuGet.MSBuild.NuGetTaskRunnerFactory" could not be loaded from the assembly
"/root/.nuget/packages/utilpack.nuget.msbuild/2.9.0/build/netcoreapp1.1/UtilPack.NuGet.MSBuild.dll". Could not load file or assembly 'NuGet.ProjectModel, Version=4.9.0.6, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.

pauldotknopf commented:
Is there anything the community can do to help?

kll commented:
I'm not aware of anything holding this up. It just seems to sit here. Pretty soon I'll need to abandon the PR because I no longer have the time to maintain it and fix all the conflicts coming from master. This needs to get merged ASAP.

kll commented:
To me that is beside the point. There is no reason this one particular thing has to wait for the fully fleshed out release process when I see commits landing in master regularly . It is creating extra work for me and I see no valid reason for it.

kll commented:
As for version numbers... I'd argue that updating to a higher major version of frameworks and SDKs qualifies as a breaking change. If we follow SemVer (which is what this repo is kinda all about) then it probably needs to be 5.0 rather than 4.1.

dazinator commented:
I agree on the breaking change semver major --> 5 as this will drop support for some lower target platforms --> so it is a breaking change.

That raises the question, what is the best way for us to bump the version in this case? Should we:-
 1. Merge this PR to master, test out the beta build (which will still be major version 4), then assuming all good tag it with stable 5.0.0? 
 2. Tag the merge to master with 5.0.0-beta.1 (i''ve not tried tagging with a pre-release label to see how gitversion handles it) so that the next beta build is on major version 5, then try that out, and then asssuming all good, tag with a stable 5.0.0?
 3. Use some other gitversion mechanism to bump the major version number like commit message or version file on this PR branch? So similar to above once PR is merged next master build is bumped to 5.0.0-beta.X and we follow up with a stable version tag.

> There is one issue we have to implement #1493, as at the moment the build scripts is not creating a release in GitHub and no release notes are generated, and there is no stable version created when we set a release tag. I started working on that, but don't have too much time to finish it. Even if we merge it now, we still have to manually trigger the stable version build.

I think the bulk of what we needed to do is done, thanks a lot to your work on the CI / CD process @arturcic . We can now release stuff to all the places, even if we have to kick of the process manually, it;s so much better than before that we actually can turn PR's arround now, by getting everything distributed to everywhere with minimal fuss. We can continue to tweak / improve the process with the release notes etc, but I don't think this should block PR's being merged any longer.

Saying that, this PR does have quite a lot of changes to review, due to the conflation of some other issues.

arturcic commented:
If we set a tag on current master as `5.0.0-alpha1` then the merged commit will have `5.0.0-beta1-1`, and later we can tag with 5.0.0

asbjornu commented:
Yep, we can tag as `5.0.0-alpha1` or we can use `next-verison: 5` in config on this branch. I'm fine with both. I also agree we should move ahead even though everything isn't perfect.

arturcic commented:
Let's merge it!

asbjornu commented:
@arturcic: Done! Let's see how this affects the build.

arturcic commented:
Great!!!, I'll check as well how the build goes

arturcic commented:
It's version 4.0.1-beta.1+53, so I suggest to wait for the build and see everything builds and than we can add 5.0.0-alpha1 as tag and do a manual build so that we get the version changed

asbjornu commented:
@arturcic: Seems like the builds succeeded:

✅ [Azure DevOps](https://dev.azure.com/GitTools/GitVersion/_build/results?buildId=88)
✅ [AppVeyor](https://ci.appveyor.com/project/GitTools/gitversion/builds/20930550)
✅ [Travis](https://travis-ci.org/GitTools/GitVersion/builds/466567845?utm_source=github_status&utm_medium=notification)

Unless anyone objects, I'll tag cf2ccfa2db2f4628259fc170cfba8de0f040c551 with `5.0.0-alpha1` so we get it out the door.

arturcic commented:
@asbjornu actually we need to fix the GitVersionTask package creation, it does not include the LibGit2Sharp for netstandard. See #1550

asbjornu commented:
@arturcic: Aha, thanks for pointing that out. That needs to be fixed, yes. We should probably also tag as `5.0.0-alpha.1` so it's consistent with GitVersion's next automatic increment `5.0.0-alpha.2`.

asbjornu commented:
@arturcic: Thanks for looking into it! 🙏
